File: Python/TextbookTrack/ba1m.py
# Implement NumberToPattern
import sys

NUCLEOTIDES = ('A', 'C', 'G', 'T')


def number_to_pattern(encoding, k):
    pattern = []
    for _ in range(k):
        encoding, remainder = divmod(encoding, 4)
        pattern.append(NUCLEOTIDES[remainder])
    return ''.join(pattern[::-1])



def main():
    data = sys.stdin.read().splitlines()
    encoding = int(data[0])
    k = int(data[1])
    result = number_to_pattern(encoding, k)
    print(result)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter01/BA1M.scala
package TextbookTrack.Chapter01

object BA1M {

  import scala.math.Integral.Implicits._

  private val Nucleotides: Vector[Char] = Vector('A', 'C', 'G', 'T')

  def numberToPattern(encoding: Int, k: Int): String = {
    val (pattern, _): (List[Char], Int) = (0 until k).foldLeft((List.empty[Char], encoding)) {
      case ((acc, code), _) =>
        val (nextCode, remainder): (Int, Int) = code /% 4
        (Nucleotides(remainder) :: acc, nextCode)
    }
    pattern.mkString
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val encoding: Int = reader.next().toInt
    val k: Int = reader.next().toInt
    val result: String = numberToPattern(encoding, k)
    println(result)
  }
}
